src/inc/til/throttled_func.h

```diff
@@ -0,0 +1,206 @@
+            }
+
+        private:
+            std::atomic<bool> _isPending;
+        };
+    } // namespace details
+
+    template<bool leading, typename... Args>
+    class throttled_func
+    {
+    public:
+        using filetime_duration = std::chrono::duration<int64_t, std::ratio<1, 10000000>>;
+        using function = std::function<void(Args...)>;
+
+        // Throttles invocations to the given `func` to not occur more often than `delay`.
+        //
+        // If this is a:
+        // * throttled_func_leading: `func` will be invoked immediately and
+        //   further invocations prevented until `delay` time has passed.
+        // * throttled_func_trailing: On the first invocation a timer of `delay` time will
+        //   be started. After the timer has expired `func` will be invoked just once.
+        //
+        // After `func` was invoked the state is reset and this cycle is repeated again.
+        throttled_func(filetime_duration delay, function func) :
+            _func{ std::move(func) },
+            _timer{ _createTimer() }
+        {
+            const auto d = -delay.count();
+            if (d >= 0)
+            {
+                throw std::invalid_argument("non-positive delay specified");
+            }
+
+            memcpy(&_delay, &d, sizeof(d));
+        }
+
+        // throttled_func uses its `this` pointer when creating _timer.
+        // Since the timer cannot be recreated, instances cannot be moved either.
+        throttled_func(const throttled_func&) = delete;
+        throttled_func& operator=(const throttled_func&) = delete;
+        throttled_func(throttled_func&&) = delete;
+        throttled_func& operator=(throttled_func&&) = delete;
+
+        // Throttles the invocation of the function passed to the constructor.
+        // If this is a trailing_throttled_func:
+        //   If you call this function again before the underlying
+        //   timer has expired, the new arguments will be used.
+        template<typename... MakeArgs>
+        void operator()(MakeArgs&&... args)
+        {
+            if (!_storage.emplace(std::forward<MakeArgs>(args)...))
+            {
+                _leading_edge();
+            }
+        }
+
+        // Modifies the pending arguments for the next function
+        // invocation, if there is one pending currently.
+        //
+        // `func` will be invoked as func(Args...). Make sure to bind any
+        // arguments in `func` by reference if you'd like to modify them.
+        template<typename F>
+        void modify_pending(F func)
+        {
+            _storage.modify_pending(func);
+        }
+
+        // Makes sure that the currently pending timer is executed
+        // as soon as possible and in that case waits for its completion.
+        // You can use this function in your destructor to ensure that any
+        // pending callback invocation is completed as soon as possible.
```

Comment:
In the original implementation:

1. If the callback was mid-firing while the caller gave up on it, it would keep itself alive and finish.
2. If the callback had been scheduled, but not called, while the caller gave up on it, it would not fire.
3. If the callback had not been scheduled, and the caller gave up on it, it would not fire.

This seems like an implementation that's different in an important way:
In cases 1 and 2, the callback will always complete. You must wait for it during destruction, and the throttled_func must remain alive for the callback's complete lifetime (scheduling + execution)...?

Reply:
Maybe I'm misunderstanding you, but the difference is:
When the new `throttled_func` is dropped/destroyed it'll cancel any timers automatically. If a callback is currently being executed in a background thread, the `throttled_func` will synchronously block in the destructor until the callback finished. It would be quite bad if you hold a mutex, while you destroy a `throttled_func`...